[SPARK-30536][CORE][SQL] Sort-merge join operator spilling performance improvements

siknezevic opened the pull request:
### What changes were proposed in this pull request?

The following list of changes will improve SQL execution performance when data is spilled on the disk:
1)	Implement lazy initialization of UnsafeSorterSpillReader - iterator on top of spilled rows:
        ... During SortMergeJoin (Left Semi Join) execution, the iterator on the spill data is created but no iteration over the data is done.
        ... Having lazy initialization of UnsafeSorterSpillReader will enable efficient processing of SortMergeJoin even if data is spilled onto disk. Unnecessary I/O will be avoided.
2)	Decrease initial memory read buffer size in UnsafeSorterSpillReader from 1MB to 1KB:
... UnsafeSorterSpillReader constructor takes lot of time due to size of default 1MB memory read buffer.
        ... The code already has logic to increase the memory read buffer if it cannot fit the data, so decreasing the size to 1K is safe and has positive performance impact.
3)	Improve memory utilization when spilling is enabled in ExternalAppendOnlyUnsafeRowArrey
        ... In the current implementation, when spilling is enabled, UnsafeExternalSorter object is created and then data moved from ExternalAppendOnlyUnsafeRowArrey object into UnsafeExternalSorter and then ExternalAppendOnlyUnsafeRowArrey object is emptied. Just before ExternalAppendOnlyUnsafeRowArrey object is emptied there are both objects in the memory with the same data. That require double memory and there is duplication of data. This can be avoided.
        ... In the proposed solution, when spark.sql.sortMergeJoinExec.buffer.in.memory.threshold is reached  adding new rows into ExternalAppendOnlyUnsafeRowArray object stops. UnsafeExternalSorter object is created and new rows are added into this object. ExternalAppendOnlyUnsafeRowArray object retains all rows already added into this object. This approach will enable better memory utilization and avoid unnecessary movement of data from one object into another.

### Why are the changes needed?

Testing with TPC-DS 100 TB benchmark data set showed that some of SQLs (example query 14) are not able to run even with extremely large Spark executor memory.Spark spilling feature has to be enabled, in order to be able to process these SQLs. Processing of SQLs becomes extremely slow when spilling is enabled. The test of this solution with query 14 and enabled spilling on the disk, showed 500X performance improvements and it didnt degrade performance of the other SQLs from TPC-DS benchmark.

The testing was done with micro-benchmark that simulates typical join scenario for 100000 rows when spilling is enabled. In this micro-benchmark the spilling will create three spill files. The test covered two scenarios: Scenario when patch is NOT applied, so micro-benchmark runs existing Spark code. Then, scenario when patch is applied. The average processing time of 100000 join rows without patch is **943989** milliseconds. The average processing time of 100000 joins rows when patch is applied is **426** milliseconds. Comparison of these two micro-benchmarks shows more than **2000 X** (943898/426 = 2215.72 X ) difference in performance. If the spilling produces more then three spill files, the performance difference would be even greater.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

By running TPC-DS SQLs with different data sets 10 TB and 100 TB
By running all Spark tests.

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

kiszk commented:
@siknezevic Thank you for your contribution. Could you please reformat the PR description?   
Which test cases can cover these new features? If not, could you please add test cases?

siknezevic commented:
@kiszk. Thank you for the questions. I hope that I reformatted PR description correctly.
There are existing tests that cover this feature. This is the list:
org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java
forcedSpillingWithReadIterator()
org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArrayBenchmark.scala
Thank you

maropu commented:
Does this change only have an good effect only on large data? For example, how about small dataset, e.g., TPCDS sf=10 and cores=1 case?

siknezevic commented:
The small data set will fit into executor memory, so there is no need to spill. I tested with just few records and it works fine. This solution will be faster for small data set when comparing with the current Spark spilling. My focus are bigger data sets and this solution will greatly improve spilling performance in those cases.

JoshRosen commented:
Regarding the "Implement lazy initialization of UnsafeSorterSpillReader" portion of the changes:

It looks like @liutang123 previously proposed a similar change in 2018 in #20184 / [SPARK-22987](https://issues.apache.org/jira/browse/SPARK-22987). I stumbled across that PR earlier this year and [left some comments](https://github.com/apache/spark/pull/20184#issuecomment-511240279): overall, I think that lazy initialization makes sense here.

In order to enable incremental review and merge of these changes, do you think it could make sense to split the lazy spill reader initialization changes into a separate PR? I haven't compared the implementations in detail, so I haven't formed any opinion on the approach here vs. in #20184.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -204,22 +195,44 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
+        result
+      } else {
+        iterator.loadNext()
+        currentRow.pointTo(iterator.getBaseObject, iterator.getBaseOffset, iterator.getRecordLength)
```
If `startIndex < numRows` then `currentRow` will be an `UnsafeRow` in `inMemoryBuffer`, so when we then roll past the `numRows` boundary then we'd start mutating that `UnsafeRow`'s state (which would corrupt the data in case we iterated over this multiple times).

As a result, I think it's safer to leave `currentRow` as it was defined before, but rename it to, say, `currentSorterRow` to make it clear that it's only used for pointing at spilled records.

siknezevic replied:
The same here. I will push all changes together. Thank you

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -106,6 +107,8 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
   def add(unsafeRow: UnsafeRow): Unit = {
     if (numRows < numRowsInMemoryBufferThreshold) {
       inMemoryBuffer += unsafeRow.copy()
+      numRows += 1
```
I think that we need to increment `numRows` and `modificationsCount` in both branches, not just this one: if we only update it in this branch then `length()` will return the wrong result.

If you want to know how many rows are in the in-memory buffer then I'd define a separate `numRowBufferedInMemory` variable; that will also help to make the code in `MergerIterator` easier to understand.

siknezevic replied:
Yes, make sense. I will push new change in the coming days. Thank you

JoshRosen reviewed (commented):
Regarding the `ExternalAppendOnlyUnsafeRowArray` change:

- It looks like there's some benchmarks for this class in `ExternalAppendOnlyUnsafeRowArrayBenchmark`; we should probably re-run those after this change.
- It seems like a lot of the complexity in `ExternalAppendOnlyUnsafeRowArray` comes from the fact that we want to define a "do not spill" threshold in terms of an absolute row count. If it wasn't for this, I think we could just directly use an `UnsafeExternalSorter` here (the performance overhead inserting a record when you don't spill should be roughly the same AFAIK, since both paths are more-or-less just copying the row's memory into the end of buffer). It looks like this was previously discussed at https://github.com/apache/spark/pull/16909#discussion_r101214678, where it sounds like the current design was motivated by performance considerations.
- Your idea of having a "merged iterator" between the in-memory and spill iterators might make sense; I left some feedback / comments on the current implementation of that idea (I think I've spotted a couple of bugs).

JoshRosen commented:
Regarding decreasing the initial memory read buffer size in `UnsafeSorterSpillReader` from 1MB to 1KB:

Intuitively this also make sense to me: AFAIK this buffer only needs to be large enough to hold a single record and records >= 1MB are probably _relatively_ rare. I'm not sure what the optimal threshold here is, though, since I don't really have data on median record sizes for representative real-world workloads.

siknezevic commented:
> Regarding the "Implement lazy initialization of UnsafeSorterSpillReader" portion of the changes:
> 
> It looks like @liutang123 previously proposed a similar change in 2018 in #20184 / [SPARK-22987](https://issues.apache.org/jira/browse/SPARK-22987). I stumbled across that PR earlier this year and [left some comments](https://github.com/apache/spark/pull/20184#issuecomment-511240279): overall, I think that lazy initialization of the individual spill readers makes sense.
> 
> In order to enable incremental review and merge of these changes, do you think it could make sense to split the lazy spill reader initialization changes into a separate PR? I haven't compared the implementations in detail, so I haven't formed any opinion on the approach here vs. the one in #20184.

Thank you for the comments.
Yes. I am OK with submitting new PR only for lazy spill reader initialization.
I looked #20184 and I can see there is still I/O in the constructor 
Line 69:  numRecords = numRecordsRemaining = dis.readInt();
But, I am open for suggestions.  Please let me know which approach would make more sense.

siknezevic commented:
> Regarding decreasing the initial memory read buffer size in `UnsafeSorterSpillReader` from 1MB to 1KB:
> 
> Intuitively this also make sense to me: AFAIK this buffer only needs to be large enough to hold a single record and records >= 1MB are probably _relatively_ rare. I'm not sure what the optimal threshold here is, though, since I don't really have data on median record sizes for representative real-world workloads.

I just want to mention that having 1 MB buffer has negative 10X impact. The constructor of UnsafeSorterSpillReader takes extra time to create 1M buffer. The benchmark tests with 1K buffer showed good performance

siknezevic commented:
> Regarding the `ExternalAppendOnlyUnsafeRowArray` change:
> 
> * It looks like there's some benchmarks for this class in `ExternalAppendOnlyUnsafeRowArrayBenchmark`; we should probably re-run those after this change.
> * It seems like a lot of the complexity in `ExternalAppendOnlyUnsafeRowArray` comes from the fact that we want to define a "do not spill" threshold in terms of an absolute row count. If it wasn't for this, I think we could just directly use an `UnsafeExternalSorter` here (the performance overhead inserting a record when you don't spill should be roughly the same AFAIK, since both paths are more-or-less just copying the row's memory into the end of buffer). It looks like this was previously discussed at [#16909 (comment)](https://github.com/apache/spark/pull/16909#discussion_r101214678), where it sounds like the current design was motivated by performance considerations.
> * Your idea of having a "merged iterator" between the in-memory and spill iterators might make sense; I left some feedback / comments on the current implementation of that idea (I think I've spotted a couple of bugs).

I agree it would be good if we could use ExternalUnsafeSorter only. ExternalAppendOnlyUnsafeRowArray class has issue that it does not have protection from OutOfMemory (OOM) like ExternalUnsafeSorter does. OOM issue (and Spark executor loss) would happen at point of doubling of capacity of the ExternalAppendOnlyUnsafeRowArray object. This object gets full and needs to expand and then OOM happens when trying to allocate memory for new “double in capacity” object (ArrayBuffer).
OOM would happen in ResizableArray.scala in ensureSize() line 103 : val newArray: Array[AnyRef] = new Array(newSize.toInt) . I have seen that over and over while testing with 100 TB data set and query 14. To avoid OOM issue you have to spill and then you run into “spilling performance” issue.

siknezevic commented:
I fixed the issues in ExternalAppendOnlyUnsafeRowArray. Next, I coming days will push new PR for lazy spill reader initialization.

siknezevic commented:
I did test to see what is the impact on the spilling performance without changes in ExternalAppendOnlyUnsafeRowArray class.The design of MergerSorterIteretor.hasNext()  in ExternalAppendOnlyUnsafeRowArray is critical for the good performance. 

The below is the stack of the code execution which has lazy initialization of the spill reader but no changes in ExternalAppendOnlyUnsafeRowArray. As you can see invocation of SpillableArrayIterator.hasNext() will cause initialization of UnsafeSorterSpillReader and it will produce I/0.

The invocation of MergerSorterIteretor.hasNext()  will not cause initialization of UnsafeSorterSpillReader, so I/0 will not happen. The startIndex that is passed to MergerSorterIteretor is 0. startIndex initialize currentIndex, so we have currentIndex < numRowBufferedInMemor (0 < numRowBufferedInMemor ) .This means that MergerSorterIteretor.hasNext()   will not invoke iterator.hasNext() method which is the case for SpillableArrayIterator.hasNext(). Invocation of iterator.hasNext() is producing I/O.

Could you please let me know if I should still submit PR for lazy spill reader initialization?
This PR without changes in ExternalAppendOnlyUnsafeRowArray class would not improve spilling performance. It would make sense only as “transition” PR that would be followed by PR for ExternalAppendOnlyUnsafeRowArray changes.


sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
org.apache.spark.io.ReadAheadInputStream.waitForAsyncReadComplete(ReadAheadInputStream.java:235)
org.apache.spark.io.ReadAheadInputStream.read(ReadAheadInputStream.java:276)
org.apache.spark.io.ReadAheadInputStream.read(ReadAheadInputStream.java:255)
java.io.DataInputStream.readInt(DataInputStream.java:387)
org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillReader.readFile(UnsafeSorterSpillReader.java:175)
org.apache.spark.util.collection.unsafe.sort.UnsafeSorterSpillReader.getNumRecords(UnsafeSorterSpillReader.java:72)
org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter$ChainedIterator.initializeNumRecords(UnsafeExternalSorter.java:733)
org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorter$ChainedIterator.hasNext(UnsafeExternalSorter.java:700)
org.apache.spark.sql.execution.ExternalAppendOnlyUnsafeRowArray$SpillableArrayIterator.hasNext(ExternalAppendOnlyUnsafeRowArray.scala:214)
org.apache.spark.sql.execution.joins.SortMergeJoinExec$$anonfun$doExecute$1$$anon$2.advanceNext(SortMergeJoinExec.scala:278)
org.apache.spark.sql.execution.RowIteratorToScala.hasNext(RowIterator.scala:68)
org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage12.agg_doAggregateWithKeys_0$(Unknown Source)
org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage12.processNext(Unknown Source)
org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
org.apache.spark.shuffle.sort.UnsafeShuffleWriter.write(UnsafeShuffleWriter.java:187)
org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:55)
org.apache.spark.scheduler.Task.run(Task.scala:123)
org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)

kiszk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -204,22 +199,41 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
+          iterator.getRecordLength)
+        currentSorterRow
+      }
+      }
```
nit: indentation issues at lines 233 and 234

shanyu commented:
Can we get this PR in please?

jiangxb1987 commented:
OK to test

siknezevic commented:
Could you please reopen this PR and remove the Stale tag? This PR brings significant performance improvements when spilling is enabled. Thank you

maropu commented:
ok to test

SparkQA commented:
**[Test build #123264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123264/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27886/
Test PASSed.

SparkQA commented:
**[Test build #123264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123264/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123264/
Test FAILed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #123269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123269/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27891/
Test PASSed.

SparkQA commented:
**[Test build #123269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123269/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123269/
Test FAILed.

gatorsmile commented:
This PR has some legit test failures. Could you fix them? 

gatorsmile commented:
cc @Ngone51 @jiangxb1987 can take a look at this?

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -703,6 +702,7 @@ public boolean hasNext() {
 
     @Override
     public void loadNext() throws IOException {
```
```
    @Override
    public void loadNext() throws IOException {
      this.hasNext();
    }
```

siknezevic replied:
Thank you for the suggestion. I will start my feedback with my understanding of your suggestion. Your idea is to reuse the code. You are proposing that content of loadNext() method be replaced by invocation of method hasNext().

I would kindly suggest if you could look loadNext() method again. Perhaps you missed the fact that last call inside of loadNext()  is “current.loadNext()”. The last call of hasNext() method is “current.hasNext()”. If we follow your recommendation, then we would change behavior of loadNext() method. I think that loadNext() moves to the next element, and hasNext() check if next element exists but it does not move to the next element. I believe that we cannot reuse the code here.

Could you please let me know if my reasoning make sense. Thank you for your time.

maropu replied:
Sounds fine.

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -720,5 +720,14 @@ public void loadNext() throws IOException {
+        }
+        numRecords += current.getNumRecords();
+      }
+    }
```
nit: How about this? (Semantically, `this.current` should be null before `initializeNumRecords` called)
```
    ChainedIterator(Queue<UnsafeSorterIterator> iterators) {
      ...
      this.current = null;
    }

    private void initializeNumRecords() throws IOException {
      if (numRecords == 0) {
        for (UnsafeSorterIterator iter: iterators) {
          numRecords += iter.getNumRecords();
        }
        this.current = iterators.remove();
      }
    }
```

siknezevic replied:
Thank you for the suggestion. It make sense. I will do it and test it.

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -47,55 +47,48 @@
   private int numRecordsRemaining;
 
-  private byte[] arr = new byte[1024 * 1024];
+  private byte[] arr = new byte[1024];
```
If this number is performance-senstive, could we parameterize it?

siknezevic replied:
Yes. I am looking into this. Thank you

siknezevic replied:
Does this look good? Perhaps you have some suggestion.

private[spark] val UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE_RATIO =
    ConfigBuilder("spark.unsafe.sorter.spill.reader.buffer.size.ratio")
      .doc("The multiplication ratio is the parameter that controls the initial read buffer " +
        "size. The multiplication ratio value range is from 1 to 1024. This parameter increases "  +
        "the initial read buffer size in 1KB increments. It will result in the initial buffer " +
        "size in the range from 1KB to 1MB. The read buffer size is dynamically adjusted " +
        "afterward based on data length read from the spilled file.")
      .intConf
      .checkValue(v => 1 <= v && v <= DEFAULT_BUFFER_SIZE_RATIO,
        s"The value must be in allowed range [1, ${DEFAULT_BUFFER_SIZE_RATIO}].")
      .createWithDefault(DEFAULT_BUFFER_SIZE_RATIO)

maropu replied:
Could you apply the change into this PR, first? Otherwise, its hard to leave comments line-by-line...

siknezevic replied:
NP. I will push update soon. I am testing changes with 10TB benchmarks.

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -148,4 +141,34 @@ public void close() throws IOException {
+  private void readFile() throws IOException {
+    assert (dataFile.length() > 0);
+    final ConfigEntry<Object> bufferSizeConfigEntry =
+            package$.MODULE$.UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE();
```
Probably, your copy-and-paste lead to wrong indents in this method. Could you check them again?

siknezevic replied:
Yah. My Intellij had 8 spaces tab. I will fix it. Thank you

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -148,4 +141,34 @@ public void close() throws IOException {
    }
   }
+
+  private void readFile() throws IOException {
```
nit: How about `readFile` -> `readSpilledFile`

siknezevic replied:
I will rename method name to readSpilledFile. Thank you

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -124,29 +129,15 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
-        }
         numFieldsPerRow = unsafeRow.numFields()
       }
-
```
nit: could you avoid the unnecessary change?

siknezevic replied:
Sure. Thank you

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -82,6 +82,8 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
 
   private var numFieldsPerRow = 0
 
+  private var numRowBufferedInMemory = 0
```
We need this variable? Could we use `inMemoryBuffer.length` instead?

siknezevic replied:
Make sense. Done. Thank you

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -168,7 +159,11 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
+                                                        startIndex - numRowBufferedInMemory
+                                                    } else 0),
+                          numFieldsPerRow,
+                          startIndex)
```
format:
```
    if (spillableArray == null) {
      new InMemoryBufferIterator(startIndex)
    } else {
      val offsetIndex = if (startIndex > numRowBufferedInMemory) {
        startIndex - numRowBufferedInMemory
      } else {
        0
      }
      new MergerIterator(
        spillableArray.getIterator(offsetIndex),
        numFieldsPerRow,
        startIndex)
    }
```

siknezevic replied:
Make sense. I will do it. Thank you

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:
```diff
@@ -204,20 +199,37 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
   }
 
-  private[this] class SpillableArrayIterator(
+  private[this] class MergerIterator(
```
nit: How about `MergerIterator` -> `SpilledArrayMergeIterator`?

siknezevic replied:
Done. Thank you

maropu commented:
Also, could you add some benchmark classes in https://github.com/apache/spark/tree/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark ?

siknezevic commented:
Thank you for the comments. I will addressed them soon

siknezevic commented:
With my next update I will address all test failures.

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #123699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123699/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28323/
Test PASSed.

SparkQA commented:
**[Test build #123699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123699/testReport)** for PR 27246 at commit [`e48a936`](https://github.com/apache/spark/commit/e48a936c0a4205e12994767ff47a21fddef60ac4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123699/
Test FAILed.

siknezevic commented:
I am aware of test failures. I will push new update that will fix test issues. Due to program logic change these tests have to be adjusted. Just need to change limits when spilling is triggered.

siknezevic commented:
> Also, could you add some benchmark classes in https://github.com/apache/spark/tree/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark ?

Hello @maropu, 
I checked available benchmarks and I can see that there is already benchmark that can be utilized.

https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCDSQueryBenchmark.scala

I am using Databrick’s Toolkit for all my testing (10TB, 100TB datasets).  TPCDSQueryBenchmark is based on Databrick’s Toolkit. I was able to test spilling with TPCDSQueryBenchmark benchmark. I executed benchmark in the following way:

/opt/spark/bin/spark-submit --class org.apache.spark.sql.execution.benchmark.TPCDSQueryBenchmark --conf 'spark.sql.sortMergeJoinExec.buffer.spill.threshold=6000' --conf 'spark.sql.sortMergeJoinExec.buffer.in.memory.threshold=1000'  '/tmp/spark-sql_2.11-2.4.6-SNAPSHOT-tests.jar' --data-location '/user/testusera1/tpcds/datasets-1g/sf1-parquet/useDecimal=true,useDate=true,filterNull=false' --query-filter 'q14a'

It runs fine and I am able to pass to it Spark config parameters to trigger spilling. I believe that we do not need new benchmark. Do you agree?

maropu commented:
`TPCDSQueryBenchmark` is for end-2-end benchmarking. I meant micro-benchmarking for the `UnsafeSorterSpillReader` class. Is it technically difficult to implement it?

SparkQA commented:
**[Test build #124141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124141/testReport)** for PR 27246 at commit [`4c78c16`](https://github.com/apache/spark/commit/4c78c166c4bc1d6aa058a1cf877ded3aa1a81995).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28758/
Test PASSed.

siknezevic commented:
> `TPCDSQueryBenchmark` is for end-2-end benchmarking. I meant micro-benchmarking for the `UnsafeSorterSpillReader` class. Is it technically difficult to implement it?

I added micro-benchmark for UnsafeSorterSpillReader into existing ExternalAppendOnlyUnsafeRowArrayBenchmark. These are results:
OpenJDK 64-Bit Server VM 1.8.0_252-8u252-b09-1~16.04-b09 on Linux 4.4.0-178-generic
Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
Spilling  SpillReader with 16000 rows:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
UnsafeSorterSpillReader_ratio1                      568            727         200          0.5        2219.4       1.0X
UnsafeSorterSpillReader_ratio1024                122902         123262         509          0.0      480085.1       0.0X

SparkQA commented:
**[Test build #124141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124141/testReport)** for PR 27246 at commit [`4c78c16`](https://github.com/apache/spark/commit/4c78c166c4bc1d6aa058a1cf877ded3aa1a81995).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124141/
Test FAILed.

SparkQA commented:
**[Test build #124142 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124142/testReport)** for PR 27246 at commit [`37d6163`](https://github.com/apache/spark/commit/37d6163b14c5bb745205f04f091e00a342a10f25).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28759/
Test PASSed.

SparkQA commented:
**[Test build #124142 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124142/testReport)** for PR 27246 at commit [`37d6163`](https://github.com/apache/spark/commit/37d6163b14c5bb745205f04f091e00a342a10f25).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124142/
Test FAILed.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArrayBenchmark.scala:
```diff
@@ -182,6 +182,76 @@ object ExternalAppendOnlyUnsafeRowArrayBenchmark extends BenchmarkBase {
+                                          numSpillThreshold: Int,
+                                          numRows: Int,
+                                          numIterators: Int,
+                                          iterations: Int): Unit = {
```
plz fix the fromat.

maropu commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -1238,6 +1239,18 @@ package object config {
+      .intConf
+      .checkValue(v => 1 <= v && v <= DEFAULT_BUFFER_SIZE_RATIO,
+        s"The value must be in allowed range [1, ${DEFAULT_BUFFER_SIZE_RATIO}].")
+      .createWithDefault(DEFAULT_BUFFER_SIZE_RATIO)
```
Could you move this config into `SQLConf`?

siknezevic replied:
Just need clarification. SQLConf is member of catalist package/project. I looked all classes involved in this change and I do not see usage of SQLConf. I would expect that new parameter UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE_RATIO is together with existing package$.MODULE$.UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE parameter  in package.scala. I just want to be sure that we want UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE_RATIO to be in SQLConf. Could you please confirm?

maropu replied:
Ah, yes. My last comment's incorrect. Your update looks okay.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArrayBenchmark.scala:
```diff
@@ -182,6 +182,76 @@ object ExternalAppendOnlyUnsafeRowArrayBenchmark extends BenchmarkBase {
+    val benchmark = new Benchmark(s"Spilling  SpillReader with $numRows rows", iterations * numRows,
+      output = output)
+
+    benchmark.addCase("UnsafeSorterSpillReader_ratio1") { _: Int =>
```
Plz update `benchmarks/ExternalAppendOnlyUnsafeRowArrayBenchmark-results.txt`.

maropu commented:
> Spilling SpillReader with 16000 rows: Best Time(ms) Avg Time(ms) Stdev(ms) Rate(M/s) Per Row(ns) Relative
UnsafeSorterSpillReader_ratio1 568 727 200 0.5 2219.4 1.0X
UnsafeSorterSpillReader_ratio1024 122902 123262 509 0.0 480085.1 0.0X

Thanks for the benchmarking! Could you run the benchmark code on both the master with/without this patch? Then, please describe performance differences in the PR description.

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -37,6 +37,10 @@
 public final class UnsafeSorterSpillReader extends UnsafeSorterIterator implements Closeable {
   public static final int MAX_BUFFER_SIZE_BYTES = 16777216; // 16 mb
+  public static final int DEFAULT_BUFFER_SIZE_RATIO = 1024;
+  private final int bufferMultiplicationRatio = SparkEnv.get() == null ? DEFAULT_BUFFER_SIZE_RATIO:
```
`SparkEnv.get()` can be null when using `UnsafeSorterSpillReader `?

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -37,6 +37,10 @@
  */
 public final class UnsafeSorterSpillReader extends UnsafeSorterIterator implements Closeable {
   public static final int MAX_BUFFER_SIZE_BYTES = 16777216; // 16 mb
+  public static final int DEFAULT_BUFFER_SIZE_RATIO = 1024;
```
Could you inline this default value in the config side (`createWithDefault(1024)`)?

maropu commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java:
```diff
@@ -47,55 +51,48 @@
   private int numRecordsRemaining;
 
-  private byte[] arr = new byte[1024 * 1024];
+  private byte[] arr = new byte[1024 * bufferMultiplicationRatio];
```
To reflect the runtime config (`spark.unsafe.sorter.spill.reader.buffer.size.ratio`), how about moving this array init into the constructor?

maropu commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -1238,6 +1239,18 @@ package object config {
       .createWithDefault(1024 * 1024)
 
+  private[spark] val UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE_RATIO =
+    ConfigBuilder("spark.unsafe.sorter.spill.reader.buffer.size.ratio")
```
Just using size for this param is a bad idea? Why do you use ratio instead of size?

siknezevic replied:
Perhaps it is my misunderstanding of your comment from 14 days ago.
You mention following:

private byte[] arr = new byte[1024 *** 1024**];
If this number is performance-senstive, could we parameterize it?

So, I created parameter. Is my understanding incorrect?

Why do you use ratio instead of size?
There is existing parameter “spark.unsafe.sorter.spill.reader.buffer.size” which is used for different purpose, so I am not able to use it here. Also, I followed your recommendation to parameterize ratio. I am OK with creating parameter for entire buffer size instead ratio. 
Please let me know what makes more sense to you.

maropu replied:
I originally meant;
```
// package.scala
  private[spark] val UNSAFE_SORTER_SPILL_READER_BUFFER_DEFAULT_SIZE_BYTES =
    ConfigBuilder("spark.unsafe.sorter.spill.reader.buffer.defaultSize")
      .internal()
      .bytesConf(ByteUnit.BYTE)
      .checkValue(...)
      .createWithDefault(1024 * 1024);

// UnsafeSorterSpillReader.java 
  private byte[] arr = null;

  public UnsafeSorterSpillReader(
      SerializerManager serializerManager,
      File file,
      BlockId blockId) throws IOException {
    assert (file.length() > 0);
    int defaultBufferSize = (int)SparkEnv.get().conf().get(package$.MODULE$.UNSAFE_SORTER_SPILL_READER_BUFFER_SIZE_BYTES())
    this.arr = new byte[DEFAULT_BUFFER_SIZE_RATIO];
```
Am I miss something?

siknezevic replied:
Thank you for the clarification. It was my misunderstanding.

siknezevic commented:
> > Spilling SpillReader with 16000 rows: Best Time(ms) Avg Time(ms) Stdev(ms) Rate(M/s) Per Row(ns) Relative
> > UnsafeSorterSpillReader_ratio1 568 727 200 0.5 2219.4 1.0X
> > UnsafeSorterSpillReader_ratio1024 122902 123262 509 0.0 480085.1 0.0X
> 
> Thanks for the benchmarking! Could you run the benchmark code on both the master with/without this patch? Then, please describe performance differences in the PR description.

@maropu
The below are results of the following micro-benchmarks:

1) Case when read buffer size is parameterized:

OpenJDK 64-Bit Server VM 1.8.0_252-8u252-b09-1~16.04-b09 on Linux 4.4.0-178-generic
Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
Spilling  SpillReader with 16000 rows:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
//------------------------------------------------------------------------------------------------
UnsafeSorterSpillReader_bufferSize1024        1312 / 1448          0.2        5125.1       1.0X
UnsafeSorterSpillReader_bufferSize1024*1024  124134 / 125262          0.0      484897.4       0.0X

2) Case when read buffer size is hard-codded to 1024 (1KB)

Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
Spilling  SpillReader with 16000 rows:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
//------------------------------------------------------------------------------------------------
UnsafeSorterSpillReader_bufferSize1024         228 /  351          1.1         892.1       1.0X

3) Case when read buffer size is hard-coded to  1024*1024 (1MB)

OpenJDK 64-Bit Server VM 1.8.0_252-8u252-b09-1~16.04-b09 on Linux 4.4.0-178-generic
Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
Spilling  SpillReader with 16000 rows:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
//------------------------------------------------------------------------------------------------
UnsafeSorterSpillReader_bufferSize1024*1024  122350 / 123832          0.0      477930.4       1.0X


If you compare the micro-benchmark results of parameterized execution 1) with hard-coded results 2) of 1024 buffer size you can noticed performance hit (around 5X). Created micro-benchmark simulates 100000 joins (creates 100000 iterators). You will notice just 2 sec performance hit for buffer 1024*1024.This is understandable because 1MB buffer size already has big performance impact.

With 10TB TPCDS data set I tested spilling with query q14a and buffer size of 1024. Execution with hard-coded read buffer size is faster by 37% (27 min vs 37 min) comparing to the execution when buffer size is parameterized and the same size 1024 is used. Query q14a, for 10TB data set, generates around 180 million joins per partition and when buffer size is parameterized, that translates into 10 min longer execution time. 

Could you please let me know would it be OK to hard-code the read buffer size to 1024? 
Taking in the consideration that UnsafeSorterSpillReader already has logic to adjust buffer size to actual data size when needed.
Thank you very much for your help

maropu commented:
> Could you please let me know would it be OK to hard-code the read buffer size to 1024?

You think the performance is independent of running platforms, e.g., CPU arch and disk I/O? If its independent, the hard-coded looks okay.

siknezevic commented:
> > Could you please let me know would it be OK to hard-code the read buffer size to 1024?
> 
> You think the performance is independent of running platforms, e.g., CPU arch and disk I/O? If its independent, the hard-coded looks okay.

Micro-benchmark was done on my dev box (single machine). 10TB benchmark was done on real 6+1 nodes (physical Ubuntu machines) cluster with fast storage and performance hit is visible in both cases. So, I think it is independent of running platforms.  I will submit new patch soon. Thank you

siknezevic commented:
> > Could you please let me know would it be OK to hard-code the read buffer size to 1024?
> 
> You think the performance is independent of running platforms, e.g., CPU arch and disk I/O? I'm not 100% sure that the `1024` value is the best on our supported platforms...
> 
> > With 10TB TPCDS data set I tested spilling with query q14a and buffer size of 1024. Execution with hard-coded read buffer size is faster by 37% (27 min vs 37 min) comparing to the execution when buffer size is parameterized and the same size 1024 is used. Query q14a, for 10TB data set, generates around 180 million joins per partition and when buffer size is parameterized, that translates into 10 min longer execution time.
> 
> Why does the parameterized one have so much overhead?

Not sure. It looks that call to package.scala to read parameter takes some time. And that time is big enough to cause performance hit because it is executed for each join row. In the case of 10TB data set there is around 180 million rows per partition.

SparkQA commented:
**[Test build #124329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124329/testReport)** for PR 27246 at commit [`0d2c0f4`](https://github.com/apache/spark/commit/0d2c0f4a2eb57496e05711cb9b66f83583d2e469).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28948/
Test PASSed.

SparkQA commented:
**[Test build #124329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124329/testReport)** for PR 27246 at commit [`0d2c0f4`](https://github.com/apache/spark/commit/0d2c0f4a2eb57496e05711cb9b66f83583d2e469).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124329/
Test FAILed.

Ngone51 commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -703,6 +702,7 @@ public boolean hasNext() {
 
     @Override
     public void loadNext() throws IOException {
+      initializeNumRecords();
       while (!current.hasNext() && !iterators.isEmpty()) {
         current = iterators.remove();
       }
```
Instead of calling `this.hasNext()`, we can extract the same lines into an inline function? e.g. `nextIterator`.

siknezevic replied:
Make sense. Thank you for the comment

Ngone51 commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -681,20 +681,19 @@ private void moveOver(UnsafeSorterIterator iter, int steps)
-      }
       this.iterators = iterators;
-      this.current = iterators.remove();
+      this.current = null;
```
this is not necessary?

siknezevic replied:
Done. Thank you

SparkQA commented:
**[Test build #124383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124383/testReport)** for PR 27246 at commit [`4c21ba6`](https://github.com/apache/spark/commit/4c21ba660fe4e992cbfaf33932abc3ce3587ebc4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/29002/
Test PASSed.

SparkQA commented:
**[Test build #124383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124383/testReport)** for PR 27246 at commit [`4c21ba6`](https://github.com/apache/spark/commit/4c21ba660fe4e992cbfaf33932abc3ce3587ebc4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124383/
Test FAILed.

siknezevic commented:
> > Spilling SpillReader with 16000 rows: Best Time(ms) Avg Time(ms) Stdev(ms) Rate(M/s) Per Row(ns) Relative
> > UnsafeSorterSpillReader_ratio1 568 727 200 0.5 2219.4 1.0X
> > UnsafeSorterSpillReader_ratio1024 122902 123262 509 0.0 480085.1 0.0X
> 
> Thanks for the benchmarking! Could you run the benchmark code on both the master with/without this patch? Then, please describe performance differences in the PR description.

I added micro-benchmark results in the section "Why are the changes needed?" of PR with description of the performance difference.

maropu commented:
Why this closed?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
> I added micro-benchmark results in the section "Why are the changes needed?" of PR with description of the performance difference.

Could you reformat the description for easy-to-read instead of just copying the benchmark result. That's because the description will be used as commit log messages.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArrayBenchmark.scala:
```diff
@@ -182,6 +182,47 @@ object ExternalAppendOnlyUnsafeRowArrayBenchmark extends BenchmarkBase {
+          0,
+          false))
+
+      for (_ <- 0L until numIterators) {
```
What's a difference with the existing benchmark code? If its almost the same, just re-running them for updating the benchmark numbers looks okay.

siknezevic replied:
No, there are not the same. Existing benchmarks read number of rows from the spilled files and iterate over the spilled records. My benchmark simulates left semi join. In that scenario there is no need to read number of records or to iterate over the spilled record. That is the reason why we have "lazy" constructor for UnsafeSorterSpillReader to avoid unnecessary reading of data from the file.  That is main point of this PR.

maropu replied:
Could you leave some comments about that here?

siknezevic replied:
During execution of sort-merge join (Left Semi Join ) for each left join row “right matches” are found and stored into ExternalAppendOnlyUnsafeRowArrey object. ExternalAppendOnlyUnsafeRowArrey object with “right matches” is created when first row on left side of the join is processed and then reused if next rows on the left side of join is the same like previous one. In the case of Queries 14a/14b there are millions of rows of “right matches” which could not fit into memory. To run this query spilling is enabled and “right matches rows” data is moved from ExternalAppendOnlyUnsafeRowArrey into UnsafeExternalSorter and then spilled onto the disk. To perform join operation on left join row, you have to create iterator on top of “right matches rows”. The operation of creation of iterator on top of “right matches” is repeated for each processed row on the left side of the join. When million rows are processed on left side of the join, the iterator on top of spilled “right matches” rows is created each time. This means that millions of time iterator on top of right matches (that are spilled on the disk) is created. The current Spark implementation creates iterator on top of spilled rows and producing I/0 because it reads number of rows stored in the spilled files but iteration action on top of iterator is never done during join operation. Iterator is created, never used and then discarded with each processed join row. This will results into millions of I/0s. One I/0 is 2 or 3 millisecond. Hence this PR which creates lazy iterator ("lazy" constructor for UnsafeSorterSpillReader), so no I/O is done. Also, my micro-benchmark simulates creation of iterators on top of spilled files which contain “right matches”. Sorry for the long explanation. Not sure if I can make it simpler. I hope it clarifies why I created micro-benchmark this way.

maropu commented on sql/core/benchmarks/ExternalAppendOnlyUnsafeRowArrayBenchmark-results.txt:
```diff
@@ -42,4 +42,8 @@ Spilling with 10000 rows:                 Best Time(ms)   Avg Time(ms)   Stdev(m
 ExternalAppendOnlyUnsafeRowArray                      9             10           1         17.1          58.5       1.2X
 
-
+OpenJDK 64-Bit Server VM 1.8.0_252-8u252-b09-1~16.04-b09 on Linux 4.4.0-178-generic
```
You mean the other numbers didn't change at all?

maropu replied:
Also, could you update the jdk11 one, too: https://github.com/apache/spark/blob/master/sql/core/benchmarks/ExternalAppendOnlyUnsafeRowArrayBenchmark-jdk11-results.txt

siknezevic replied:
> You mean the other numbers didn't change at all?

My understanding was to add new benchmark results to the file. I didn't change other results in the files. Do you want me to update all results?

siknezevic replied:
One more thing, we will not see performance improvements for the existing benchmarks if the benchmark iterates  over the spilled data. Then comparison will be related to the machine on which benchmark was running and disk speed.

maropu replied:
> My understanding was to add new benchmark results to the file. I didn't change other results in the files. Do you want me to update all results?

We need to check no perf regression on jdk8/jdk11

siknezevic replied:
Yes, I did comparison of the existing micro-benchmark results when patch is applied and when there is no patch.
I do not see regression in the performance for the existing micro-benchmarks. I will push micro-benchmark results for jdk11 soon.
Thank you for help 😊

siknezevic commented:
> Why this closed?

Sorry, I added info about what I did and I thought the task is done. Apparently I was wrong

siknezevic commented:
> > I added micro-benchmark results in the section "Why are the changes needed?" of PR with description of the performance difference.
> 
> Could you reformat the description for easy-to-read instead of just copying the benchmark result. That's because the description will be used as commit log messages.

I changed the description. Could you please review it? Thank you for the help !!

maropu commented on sql/core/benchmarks/ExternalAppendOnlyUnsafeRowArrayBenchmark-results.txt:
```diff
@@ -42,4 +42,8 @@ Spilling with 10000 rows:                 Best Time(ms)   Avg Time(ms)   Stdev(m
+Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
+Spilling  SpillReader with 16000 rows:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+UnsafeSorterSpillReader_bufferSize1024              411            426          13          0.6        1607.2       1.0X
```
What's a number without this PR?

siknezevic replied:
Here it is:
OpenJDK 64-Bit Server VM 1.8.0_252-8u252-b09-1~16.04-b09 on Linux 4.4.0-178-generic
Intel(R) Xeon(R) CPU E5-2687W v3 @ 3.10GHz
Spilling  SpillReader with 16000 rows:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
//------------------------------------------------------------------------------------------------------------------------
UnsafeSorterSpillReader                          932683         943898         NaN          0.0     3643292.4       1.0X

siknezevic replied:
There are three spill files. It will result in three I/Os for each created iterator, because it has to read number of rows from each spilled file. So, we see more than 2000 X performance difference.

maropu replied:
oh, looks nice.

maropu commented:
Looks okay to me. Anyone could check this? @cloud-fan @dongjoon-hyun  @JoshRosen @jiangxb1987

siknezevic commented:
Just checking to see if there is anything on my side that I should do to move this PR. Thank you

maropu commented:
kindly ping: @cloud-fan @dongjoon-hyun @JoshRosen @jiangxb1987

gatorsmile commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31484/
Test PASSed.

SparkQA commented:
**[Test build #126860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126860/testReport)** for PR 27246 at commit [`2e91826`](https://github.com/apache/spark/commit/2e91826832d24ea25384bca1f860c2fe68b4fc5c).

SparkQA commented:
**[Test build #126860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126860/testReport)** for PR 27246 at commit [`2e91826`](https://github.com/apache/spark/commit/2e91826832d24ea25384bca1f860c2fe68b4fc5c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126860/
Test FAILed.

SparkQA commented:
**[Test build #126900 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126900/testReport)** for PR 27246 at commit [`b824828`](https://github.com/apache/spark/commit/b824828e732017d4c1eda470704b9e7a53cdad95).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31525/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126900/
Test FAILed.

maropu commented:
@siknezevic Seems like a valid test failure, so could you check it? https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126900/testReport/org.apache.spark.sql/JoinSuite/test_SortMergeJoin__with_spill_/

siknezevic committed:
Sort-merge join operator spilling performance improvements

What changes were proposed in this pull request?

The following list of changes will improve SQL execution performance when data is spilled on the disk:
1)	Implement lazy initialization of UnsafeSorterSpillReader - iterator on top of spilled rows:
... During SortMergeJoin (Left Semi Join) execution, the iterator on the spill data is created but no iteration over the data is done.
... Having lazy initialization of UnsafeSorterSpillReader will enable efficient processing of SortMergeJoin even if data is spilled onto disk. Unnecessary I/O will be avoided.
2)	Decrease initial memory read buffer size in UnsafeSorterSpillReader from 1MB to 1KB:
... UnsafeSorterSpillReader constructor takes lot of time due to size of default 1MB memory read buffer.
... The code already has logic to increase the memory read buffer if it cannot fit the data, so decreasing the size to 1K is safe and has positive performance impact.
3)	Improve memory utilization when spilling is enabled in ExternalAppendOnlyUnsafeRowArrey
... In the current implementation, when spilling is enabled, UnsafeExternalSorter object is created and then data moved from ExternalAppendOnlyUnsafeRowArrey object into UnsafeExternalSorter and then ExternalAppendOnlyUnsafeRowArrey object is emptied. Just before ExternalAppendOnlyUnsafeRowArrey object is emptied there are both objects in the memory with the same data. That require double memory and there is duplication of data. This can be avoided.
... In the proposed solution, when spark.sql.sortMergeJoinExec.buffer.in.memory.threshold is reached  adding new rows into ExternalAppendOnlyUnsafeRowArray object stops. UnsafeExternalSorter object is created and new rows are added into this object. ExternalAppendOnlyUnsafeRowArray object retains all rows already added into this object. This approach will enable better memory utilization and avoid unnecessary movement of data from one object into another.

Why are the changes needed?

Testing with TPC-DS 100 TB benchmark data set showed that some of SQLs (example query 14) are not able to run even with extremely large Spark executor memory.
Spark spilling feature has to be enabled, in order to be able to process these SQLs. Processing of SQLs becomes extremely slow when spilling is enabled.
The test of this solution with query 14 and enabled spilling on the disk, showed 500X performance improvements and it didnt degrade performance of the other SQLs from TPC-DS benchmark.

Does this PR introduce any user-facing change?

No

How was this patch tested?

By running TPC-DS SQLs with different data sets 10 TB and 100 TB
By running all Spark tests.

SparkQA commented:
**[Test build #126922 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126922/testReport)** for PR 27246 at commit [`9d5bd99`](https://github.com/apache/spark/commit/9d5bd99e317d5d0f73b645f8e29276c5d99ed39a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31546/
Test PASSed.

SparkQA commented:
**[Test build #126922 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126922/testReport)** for PR 27246 at commit [`9d5bd99`](https://github.com/apache/spark/commit/9d5bd99e317d5d0f73b645f8e29276c5d99ed39a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126922/
Test PASSed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #128448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128448/testReport)** for PR 27246 at commit [`9d5bd99`](https://github.com/apache/spark/commit/9d5bd99e317d5d0f73b645f8e29276c5d99ed39a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33072/
Test PASSed.

SparkQA commented:
**[Test build #128448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128448/testReport)** for PR 27246 at commit [`9d5bd99`](https://github.com/apache/spark/commit/9d5bd99e317d5d0f73b645f8e29276c5d99ed39a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128448/
Test PASSed.

caican00 commented:
@siknezevic  Hi, is there any progress on this pr?

siknezevic commented:
No.
